File: pacman/analyze/prediction.py
"""Predict bounded ghost movement while respecting current direction."""

from dataclasses import dataclass
from enum import Enum

from typed_errs import Err, Ok, Result

from pacman.analyze.models import MazeGraph, Move
from pacman.analyze.threat import NO_THREAT, ThreatField
from pacman.replay.models import Direction, Ghost, GhostFrame, GhostState, Maze, TileIndex

OPPOSITE: dict[Direction, Direction] = {
    Direction.UP: Direction.DOWN,
    Direction.RIGHT: Direction.LEFT,
    Direction.DOWN: Direction.UP,
    Direction.LEFT: Direction.RIGHT,
}


class PredictionError(Enum):
    """Failures encountered while predicting ghost movement."""

    INVALID_GHOST_TILE = "invalid_ghost_tile"
    INVALID_HORIZON = "invalid_horizon"


@dataclass(frozen=True, slots=True)
class PredictedGhostState:
    """One possible ghost state at a future relative tick."""

    ghost: Ghost
    tile: TileIndex
    direction: Direction
    tick: int
    dangerous: bool


@dataclass(frozen=True, slots=True)
class GhostPrediction:
    """Possible states grouped by relative prediction tick."""

    ghost: Ghost
    ticks: tuple[tuple[PredictedGhostState, ...], ...]


def prediction_err(error: PredictionError) -> Err[PredictionError]:
    """Create a prediction error with consistent context.

    Returns:
        A contextual prediction error.
    """
    return Err(
        error=error,
        namespace="prediction",
        context_msg="Failed to predict ghost movement",
    )


def legal_ghost_moves(
    graph: MazeGraph,
    tile: TileIndex,
    direction: Direction,
) -> tuple[Move, ...]:
    """Return moves excluding reversal unless reversal is the only choice."""
    moves = graph.neighbors(tile)
    forward = tuple(move for move in moves if move.direction is not OPPOSITE[direction])
    return forward if forward else moves


def predict_ghost(
    graph: MazeGraph,
    maze: Maze,
    ghost: GhostFrame,
    horizon: int,
) -> Result[GhostPrediction, PredictionError]:
    """Enumerate possible direction-constrained ghost states through a horizon.

    Returns:
        A bounded immutable prediction or a typed validation error.
    """
    if horizon < 0:
        return prediction_err(PredictionError.INVALID_HORIZON)
    origin = maze.tile_index(ghost.position)
    if not graph.contains(origin):
        return prediction_err(PredictionError.INVALID_GHOST_TILE)

    dangerous = ghost.state not in (GhostState.FRIGHTENED, GhostState.EATEN)
    current = (PredictedGhostState(ghost.ghost, origin, ghost.direction, 0, dangerous),)
    ticks: list[tuple[PredictedGhostState, ...]] = [current]

    for tick in range(1, horizon + 1):
        next_states: dict[tuple[TileIndex, Direction], PredictedGhostState] = {}
        for state in current:
            for move in legal_ghost_moves(graph, state.tile, state.direction):
                predicted = PredictedGhostState(
                    ghost=state.ghost,
                    tile=move.destination,
                    direction=move.direction,
                    tick=tick,
                    dangerous=state.dangerous,
                )
                next_states[(predicted.tile, predicted.direction)] = predicted
        current = tuple(next_states.values())
        ticks.append(current)

    return Ok(GhostPrediction(ghost.ghost, tuple(ticks)))


def build_predicted_threat_field(
    graph: MazeGraph,
    maze: Maze,
    ghosts: tuple[GhostFrame, ...],
    horizon: int,
) -> Result[ThreatField, PredictionError]:
    """Combine earliest dangerous arrivals from bounded ghost predictions.

    Returns:
        A threat field or a typed prediction error.
    """
    if horizon < 0:
        return prediction_err(PredictionError.INVALID_HORIZON)

    etas = [NO_THREAT] * len(graph.moves)
    owners: list[list[Ghost]] = [[] for _ in graph.moves]
    for ghost in ghosts:
        prediction = predict_ghost(graph, maze, ghost, horizon)
        if isinstance(prediction, Err):
            return prediction
        for states in prediction.value.ticks:
            for state in states:
                if not state.dangerous:
                    continue
                index = int(state.tile)
                if etas[index] == NO_THREAT or state.tick < etas[index]:
                    etas[index] = state.tick
                    owners[index] = [state.ghost]
                elif state.tick == etas[index] and state.ghost not in owners[index]:
                    owners[index].append(state.ghost)

    return Ok(ThreatField(tuple(etas), tuple(tuple(value) for value in owners)))

File: native/tests/prediction.cpp
#include <array>
#include <cstdint>

#include <pacman/macros.h>

import pacman.graph;
import pacman.prediction;
import pacman.types;

namespace {

struct prediction_fixture {
    std::array<std::uint32_t, 3> arrivals{};
    std::array<pacman::prediction_state, 12> current{};
    std::array<pacman::prediction_state, 12> next{};
    std::array<std::uint32_t, 12> seen{};

    fn workspace() noexcept -> pacman::prediction_workspace {
        return {
            .earliest_arrival = arrivals,
            .current = current,
            .next = next,
            .seen = seen,
        };
    }
};

fn corridor() -> std::array<pacman::tile_neighbors, 3> {
    using enum pacman::direction;
    return {{
        {.moves = {{{.destination = 1, .heading = right}}}, .count = 1},
        {.moves = {{{.destination = 0, .heading = left},
                    {.destination = 2, .heading = right}}},
         .count = 2},
        {.moves = {{{.destination = 1, .heading = left}}}, .count = 1},
    }};
}

fn predict(const std::array<pacman::tile_neighbors, 3> &tiles,
           const pacman::tile_index tile, const pacman::direction heading,
           prediction_fixture &fixture) -> bool {
    return pacman::predict_ghost(
        {.tiles = tiles},
        {.tile = tile, .heading = heading, .ghost = 0, .dangerous = true}, 1,
        fixture.workspace());
}

fn test_continues_forward() -> bool {
    const let tiles = corridor();
    prediction_fixture fixture{};

    return predict(tiles, 1, pacman::direction::right, fixture) &&
           fixture.arrivals[0] == pacman::no_prediction &&
           fixture.arrivals[1] == 0 && fixture.arrivals[2] == 1;
}

fn test_reverses_at_dead_end() -> bool {
    const let tiles = corridor();
    prediction_fixture fixture{};

    return predict(tiles, 2, pacman::direction::right, fixture) &&
           fixture.arrivals[1] == 1 &&
           fixture.next[0].heading == pacman::direction::left;
}

fn test_uses_wraparound_edge() -> bool {
    using enum pacman::direction;
    const std::array<pacman::tile_neighbors, 3> tiles{{
        {.moves = {{{.destination = 1, .heading = right},
                    {.destination = 2, .heading = left, .wraparound = true}}},
         .count = 2},
        {.moves = {{{.destination = 0, .heading = left},
                    {.destination = 2, .heading = right}}},
         .count = 2},
        {.moves = {{{.destination = 1, .heading = left},
                    {.destination = 0, .heading = right, .wraparound = true}}},
         .count = 2},
    }};
    prediction_fixture fixture{};

    return predict(tiles, 0, left, fixture) && fixture.arrivals[2] == 1;
}

fn test_combines_predicted_threats() -> bool {
    const let tiles = corridor();
    prediction_fixture fixture{};
    std::array<std::uint32_t, 3> eta{};
    std::array<std::uint8_t, 3> owners{};
    const std::array ghosts{
        pacman::ghost_prediction_input{
            .tile = 0,
            .heading = pacman::direction::right,
            .ghost = 0,
            .dangerous = true,
        },
        pacman::ghost_prediction_input{
            .tile = 2,
            .heading = pacman::direction::left,
            .ghost = 1,
            .dangerous = true,
        },
    };

    return pacman::build_predicted_threat_field(
               {.tiles = tiles}, ghosts, 1, fixture.workspace(),
               {.eta = eta, .owners = owners}) &&
           eta == std::array<std::uint32_t, 3>{0, 1, 0} &&
           owners == std::array<std::uint8_t, 3>{1, 3, 2};
}

} // namespace

fn run_prediction_tests() -> int {
    if (!test_continues_forward()) {
        return 20;
    }
    if (!test_reverses_at_dead_end()) {
        return 21;
    }
    if (!test_uses_wraparound_edge()) {
        return 22;
    }
    if (!test_combines_predicted_threats()) {
        return 23;
    }
    return 0;
}
